Mathlib/Topology/Order/Monotone.lean

```diff
@@ -44,6 +44,125 @@ lemma MonotoneOn.insert_of_continuousWithinAt [TopologicalSpace β] [OrderClosed
     filter_upwards [this] with y hy
     exact hf hy.1 hb (le_of_lt hy.2)
 
+/-- If a function is monotone on a set in a second countable topological space, then there
+are only countably many points that have several preimages. -/
+lemma MonotoneOn.countable_setOf_two_preimages [SecondCountableTopology α]
+    (hf : MonotoneOn f s) :
+    Set.Countable {c | ∃ x y, x ∈ s ∧ y ∈ s ∧ x < y ∧ f x = c ∧ f y = c} := by
+  nontriviality α
+  let t := {c | ∃ x, ∃ y, x ∈ s ∧ y ∈ s ∧ x < y ∧ f x = c ∧ f y = c}
+  have : ∀ c ∈ t, ∃ x, ∃ y, x ∈ s ∧ y ∈ s ∧ x < y ∧ f x = c ∧ f y = c := fun c hc ↦ hc
+  choose! x y hxs hys hxy hfx hfy using this
+  let u := x '' t
+  suffices H : Set.Countable (x '' t) by
+    have : Set.InjOn x t := by
+      intro c hc d hd hcd
+      have : f (x c) = f (x d) := by simp [hcd]
+      rwa [hfx _ hc, hfx _ hd] at this
+    exact countable_of_injective_of_countable_image this H
+  apply Set.PairwiseDisjoint.countable_of_Ioo (y := fun a ↦ y (f a)); swap
+  · rintro a ⟨c, hc, rfl⟩
+    rw [hfx _ hc]
+    exact hxy _ hc
+  simp only [PairwiseDisjoint, Set.Pairwise, mem_image, onFun, forall_exists_index, and_imp,
+    forall_apply_eq_imp_iff₂]
+  intro c hc d hd hcd
+  wlog H : c < d generalizing c d with h
+  · apply (h d hd c hc hcd.symm ?_).symm
+    have : c ≠ d := fun h ↦ hcd (congrArg x h)
+    exact lt_of_le_of_ne (not_lt.1 H) this.symm
```

Comment:
Alternatively; import `Mathlib.Tactic.Order` and solve this goal by `order`: optional